### PR TITLE
sql: optimize SRF rewriting to avoid cross joins in simple cases

### LINKED ISSUE #17546: sql: SELECT * FROM generate_series() vs SELECT FROM generate_series()
I was surprised with @tso to find out that there's a difference between the two. The 2nd one doesn't seem to produce results until the function terminates producing the whole result table, whereas the first one "streams" results out. Is that expected?

### PR DESCRIPTION
**Pull request by knz**
tl;dr: make a common case faster and use less memory.

Set-generating functions (SRFs) are meant to be used as operand to a
FROM clause, for example `SELECT * FROM generate_series(1, 10)`.

However, PostgreSQL also supports the degenerate syntax

```sql
SELECT generate_series() FROM x, y, z
```

which (somewhat counter-intuitively, but oh well) really means:

```sql
SELECT generate_series FROM x, y, z, generate_series() as unnamed(generate_series)
```

In other words, a SRF present in the render part of a SELECT clause is
interpreted as the source of a cross-join with the existing FROM
clause. This recursively extends to multiple SRFs, for example:

```sql
SELECT generate_series(1,2), generate_series(2,3) FROM x, y, z
-- is really, in disguise:
SELECT a.generate_series, b.generate_series
  FROM x, y, z,
       generate_series(1,2) AS a(generate_series),
       generate_series(2,3) AS b(generate_series)
```

An interesting edge case, unfortunately rather common in interactive
shells, is the case of `SELECT generate_series()` *without a FROM
clause*. This is common because SRFs look like functions and users
when exploring will try out "functions" in that way.

Now, what does a SELECT clause mean without a FROM clause?

From a relational-algebraic perspective, "SELECT X FROM Y" means "for
every row in Y, render X". Now, if there is no FROM clause, we still
see a result. How can that be? This is because SQL mandates that a
missing FROM clause is substituted by *a virtual data source with no
columns and exactly one row*. This is called the "unary source".

In CockroachDB we use `emptyNode{results:true}` to represent this.

Therefore, the query `SELECT generate_series()` really means
`SELECT ... FROM <unary> CROSS JOIN generate_series()`.

Without further consideration, this is semantically correct and
produces valid results: the result of a cross join between the unary
source with any source Y returns the same rows as Y.

The issue that this patch addresses is that this particular flavor of
join (with a SRF operand) is not distributed and executed using an
in-memory hash join. This in turns requires accumulating all the
rows from the RHS of the join in memory before the join "resolves"
and sees that all these rows can be emitted as-is as a result.

This behavior, combined with the announcement of a prior patch that
makes CockroachDB "stream results" during execution towards the
client, creates a possible user surprise: despite the fact CockroachDB
streams results, a "simple query" like `SELECT
generate_series(1,10000000)` would not be streamed (and possibly cause
memory exhaustion), while the more canonical `SELECT * FROM
generate_series(1,10000000)` would be streamed properly (because it
doesn't hide a cross-join in disguise).

To address this possible surprise, this patch special-cases the
expansion of a SRF in render position to a cross-join, when the
existing FROM clause is the unary source. In that case, the SRF is
used as-is and no join is generated.

Idea courtesy of @jordanlewis + @justinj.

Fixes  #17546.

### CODE REVIEW

**Comment by justinj** [comment]
LGTM!

---

Review status: 0 of 2 files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---

*[pkg/sql/render.go, line 455 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/17835#-KsETH08dD4c0Oxq741g:-KsETH08dD4c0Oxq741h:b-a06r6w) ([raw file](https://github.com/cockroachdb/cockroach/blob/d974d819eaaa87699fe79834d8431990739afc4a/pkg/sql/render.go#L455)):*
> ```Go
> // a pseudo-table with zero columns and exactly one row.
> func isUnarySource(src planDataSource) bool {
> 	e, ok := src.plan.(*emptyNode)
> ```

Is there a reason `emptyNode`s are used for sources both with 0 rows and with 1 row? Might make this a little more locally understandable if this were just `_, ok := src.plan.(*unaryNode); return ok` (or if we wanted to be cute it could be like, `joinIdentityNode`), and then we also had `zeroNode`, or something. I had to look up the definition of `emptyNode` to see what `results` meant.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/17835#-:-KsEXlLNenvAPagKcegc:b-85i648)*

**Comment by knz** [comment]
TFYR! forked the renaming to #17839. :)
